src/applications/apps.js

```diff
@@ -228,3 +201,172 @@ function immediatelyUnloadApp(app, resolve, reject) {
     })
     .catch(reject);
 }
+
+function validateRegisterWithArguments(
+  name,
+  appOrLoadApp,
+  activeWhen,
+  customProps
+) {
+  if (typeof name !== "string" || name.length === 0)
+    throw Error(
+      formatErrorMessage(
+        20,
+        __DEV__ &&
+          `The 1st argument to registerApplication must be a non-empty string 'appName'`
+      )
+    );
+
+  if (!appOrLoadApp)
+    throw Error(
+      formatErrorMessage(
+        23,
+        __DEV__ &&
+          "The 2nd argument to registerApplication must be an application or loading application function"
+      )
+    );
+
+  if (typeof activeWhen !== "function")
+    throw Error(
+      formatErrorMessage(
+        24,
+        __DEV__ &&
+          "The 3rd argument to registerApplication must be an activeWhen function"
+      )
+    );
+
+  if (
+    !!customProps &&
+    (typeof customProps !== "object" || Array.isArray(customProps))
+  )
+    throw Error(
+      formatErrorMessage(
+        22,
+        __DEV__ &&
+          "The optional 4th argument is a customProps and must be an object"
+      )
+    );
+}
+
+export function validateRegisterWithConfig(config) {
+  if (Array.isArray(config) || config === null)
+    throw Error(
+      formatErrorMessage(
+        31,
+        __DEV__ && "Configuration object can't be an Array or null!"
+      )
+    );
+
+  const validKeys = ["name", "app", "activeWhen", "customProps"];
+
+  const invalidKeys = Object.keys(config).reduce(
+    (invalidKeys, prop) =>
+      validKeys.includes(prop) ? invalidKeys : invalidKeys.concat(prop),
+    []
+  );
+
+  if (invalidKeys.length !== 0)
+    throw Error(
+      formatErrorMessage(
+        30,
+        __DEV__ &&
+          `The configuration object accepts only: ${validKeys.join(
+            ", "
+          )}. Invalid keys: ${invalidKeys.join(", ")}.`
+      )
+    );
+
+  if (typeof config.name !== "string" || config.name.length === 0)
+    throw Error(
+      formatErrorMessage(
+        20,
+        __DEV__ &&
+          "The config.name on registerApplication must be a non-empty string"
+      )
+    );
+
+  if (!config.app)
+    throw Error(
+      formatErrorMessage(
+        20,
+        __DEV__ &&
+          "The config.app on registerApplication must be an application or a loading fuinction"
+      )
+    );
+
+  // @TODO this is not yet implemented
```

Comment:
Let's save this for your next PR, and not include it in this one. You can move this to a different branch for now and then once this PR is merged you can use that branch for your next PR.